File: fashion_navya/fashion_navya/report/user_warehouse_stock_balance/user_warehouse_stock_balance.py
# # # Copyright (c) 2026, pawasthy11@gmail.com and contributors
# # # For license information, please see license.txt

# # # import frappe


# # def execute(filters=None):
# # 	columns, data = [], []
# # 	return columns, data
































# import frappe
# from frappe import _
# from frappe.utils import getdate
# from erpnext.stock.report.stock_balance.stock_balance import StockBalanceReport


# def execute(filters=None):

#     filters = frappe._dict(filters or {})

#     # ---------------- VALIDATION ----------------
#     if not filters.get("user"):
#         frappe.throw(_("Please select User"))

#     if not filters.get("from_date"):
#         frappe.throw(_("From Date is required"))

#     if not filters.get("to_date"):
#         frappe.throw(_("To Date is required"))

#     filters.from_date = getdate(filters.from_date)
#     filters.to_date = getdate(filters.to_date)

#     # ---------------- USER WAREHOUSES ----------------
#     warehouses = frappe.get_all(
#         "Location Wise Warehoue",
#         filters={"user": filters.user},
#         pluck="warehouse"
#     )

#     warehouses = sorted({w for w in warehouses if w})

#     if not warehouses:
#         return get_columns([]), []

#     # ---------------- IMPORTANT FIX ----------------
#     # DO NOT pass warehouse into StockBalanceReport
#     filters.pop("warehouse", None)

#     # ---------------- RUN REPORT ----------------
#     stock_report = StockBalanceReport(filters)
#     columns, stock_data = stock_report.run()

#     # ---------------- FILTER AFTER ----------------
#     filtered_data = [
#         row for row in stock_data
#         if row.get("warehouse") in warehouses
#     ]

#     # ---------------- PIVOT ----------------
#     item_map = {}

#     for row in filtered_data:

#         item_code = row.get("item_code")
#         warehouse = row.get("warehouse")
#         qty = row.get("bal_qty") or 0

#         if not item_code or not warehouse:
#             continue

#         if item_code not in item_map:

#             item_map[item_code] = {
#                 "user": filters.user,
#                 "item_code": item_code,
#                 "item_name": row.get("item_name")
#             }

#             for wh in warehouses:
#                 item_map[item_code][frappe.scrub(wh)] = 0

#         item_map[item_code][frappe.scrub(warehouse)] = qty

#     return get_columns(warehouses), list(item_map.values())


# def get_columns(warehouses):

#     columns = [
#         {
#             "label": _("User"),
#             "fieldname": "user",
#             "fieldtype": "Link",
#             "options": "User",
#             "width": 180
#         },
#         {
#             "label": _("Item Code"),
#             "fieldname": "item_code",
#             "fieldtype": "Link",
#             "options": "Item",
#             "width": 180
#         },
#         {
#             "label": _("Item Name"),
#             "fieldname": "item_name",
#             "fieldtype": "Data",
#             "width": 250
#         }
#     ]

#     for wh in warehouses:
#         columns.append({
#             "label": wh,
#             "fieldname": frappe.scrub(wh),
#             "fieldtype": "Float",
#             "width": 140
#         })

#     return columns



















































# ##################################################
# import frappe
# from frappe import _
# from frappe.utils import getdate
# from erpnext.stock.report.stock_balance.stock_balance import StockBalanceReport


# def execute(filters=None):

#     filters = frappe._dict(filters or {})

#     # ---------------- VALIDATION ----------------

#     if not filters.get("user"):
#         frappe.throw(_("Please select User"))

#     if not filters.get("from_date"):
#         frappe.throw(_("From Date is required"))

#     if not filters.get("to_date"):
#         frappe.throw(_("To Date is required"))

#     filters.from_date = getdate(filters.from_date)
#     filters.to_date = getdate(filters.to_date)

#     # ---------------- MULTISELECT WAREHOUSE ----------------

#     warehouses = filters.get("warehouse")

#     if not warehouses:
#         frappe.throw(_("Please select Warehouse"))

#     # MultiSelectList returns comma separated string
#     if isinstance(warehouses, str):

#         warehouses = [
#             d.strip()
#             for d in warehouses.split(",")
#             if d.strip()
#         ]

#     warehouses = sorted(set(warehouses))

#     if not warehouses:
#         return get_columns([]), []

#     # ---------------- IMPORTANT FIX ----------------
#     # DO NOT pass warehouse directly
#     filters.pop("warehouse", None)

#     # ---------------- RUN STOCK REPORT ----------------

#     stock_report = StockBalanceReport(filters)

#     columns, stock_data = stock_report.run()

#     # ---------------- FILTER SELECTED WAREHOUSES ----------------

#     filtered_data = [

#         row for row in stock_data

#         if row.get("warehouse") in warehouses

#     ]

#     # ---------------- PIVOT ----------------

#     item_map = {}

#     for row in filtered_data:

#         item_code = row.get("item_code")
#         warehouse = row.get("warehouse")
#         qty = row.get("bal_qty") or 0

#         if not item_code or not warehouse:
#             continue

#         if item_code not in item_map:

#             item_map[item_code] = {

#                 "user": filters.user,
#                 "item_code": item_code,
#                 "item_name": row.get("item_name")

#             }

#             # Initialize all selected warehouses
#             for wh in warehouses:

#                 item_map[item_code][
#                     frappe.scrub(wh)
#                 ] = 0

#         item_map[item_code][
#             frappe.scrub(warehouse)
#         ] = qty

#     data = list(item_map.values())

#     return get_columns(warehouses), data


# def get_columns(warehouses):

#     columns = [

#         {
#             "label": _("User"),
#             "fieldname": "user",
#             "fieldtype": "Link",
#             "options": "User",
#             "width": 180
#         },

#         {
#             "label": _("Item Code"),
#             "fieldname": "item_code",
#             "fieldtype": "Link",
#             "options": "Item",
#             "width": 180
#         },

#         {
#             "label": _("Item Name"),
#             "fieldname": "item_name",
#             "fieldtype": "Data",
#             "width": 250
#         }

#     ]

#     # Dynamic warehouse columns
#     for wh in warehouses:

#         columns.append({

#             "label": wh,
#             "fieldname": frappe.scrub(wh),
#             "fieldtype": "Float",
#             "width": 140

#         })

#     return columns
# ##############################################################################    # 















######### fashion_navya.fashion_navya.report.user_warehouse_stock_balance.user_warehouse_stock_balance.get_user_warehouses

import frappe
from frappe import _
from frappe.utils import getdate
from erpnext.stock.report.stock_balance.stock_balance import StockBalanceReport


@frappe.whitelist()
def get_user_warehouses(user):

    warehouses = frappe.get_all(
        "Location Wise Warehoue",
        filters={
            "parenttype": "Permitted Files",
            "user": user
        },
        fields=["warehouse"]
    )

    unique_warehouses = sorted(
        list(
            {
                d.warehouse
                for d in warehouses
                if d.warehouse
            }
        )
    )

    return unique_warehouses


def execute(filters=None):

    filters = frappe._dict(filters or {})

    if not filters.get("user"):
        frappe.throw(_("Please select User"))

    if not filters.get("from_date"):
        frappe.throw(_("From Date is required"))

    if not filters.get("to_date"):
        frappe.throw(_("To Date is required"))

    filters.from_date = getdate(filters.from_date)
    filters.to_date = getdate(filters.to_date)

    warehouses = filters.get("warehouse")

    if not warehouses:

        warehouses = get_user_warehouses(
            filters.user
        )

    if isinstance(warehouses, str):

        warehouses = [
            d.strip()
            for d in warehouses.split(",")
            if d.strip()
        ]

    warehouses = sorted(set(warehouses))

    if not warehouses:
        return get_columns([]), []

    filters.pop("warehouse", None)

    stock_report = StockBalanceReport(filters)

    columns, stock_data = stock_report.run()

    filtered_data = [

        row for row in stock_data

        if row.get("warehouse") in warehouses

    ]

    item_map = {}

    for row in filtered_data:

        item_code = row.get("item_code")
        warehouse = row.get("warehouse")
        qty = row.get("bal_qty") or 0

        if not item_code or not warehouse:
            continue

        if item_code not in item_map:

            item_map[item_code] = {

                "user": filters.user,
                "item_code": item_code,
                "item_name": row.get("item_name")

            }

            for wh in warehouses:

                item_map[item_code][
                    frappe.scrub(wh)
                ] = 0

        item_map[item_code][
            frappe.scrub(warehouse)
        ] = qty

    data = list(item_map.values())

    return get_columns(warehouses), data


def get_columns(warehouses):

    columns = [

        {
            "label": _("User"),
            "fieldname": "user",
            "fieldtype": "Link",
            "options": "User",
            "width": 180
        },

        {
            "label": _("Item Code"),
            "fieldname": "item_code",
            "fieldtype": "Link",
            "options": "Item",
            "width": 180
        },

        {
            "label": _("Item Name"),
            "fieldname": "item_name",
            "fieldtype": "Data",
            "width": 250
        }

    ]

    for wh in warehouses:

        columns.append({

            "label": wh,
            "fieldname": frappe.scrub(wh),
            "fieldtype": "Float",
            "width": 140

        })

    return columns

File: fashion_navya/fashion_navya/report/user_warehouse_stock_balance/user_warehouse_stock_balance.js
// // // Copyright (c) 2026, pawasthy11@gmail.com and contributors
// // // For license information, please see license.txt

// // frappe.query_reports["User Warehouse Stock Balance"] = {
// // 	"filters": [

// // 	]
// // };














// frappe.query_reports["User Warehouse Stock Balance"] = {

//     filters: [

//         {
//             fieldname: "user",
//             label: __("User"),
//             fieldtype: "Link",
//             options: "User",
//             reqd: 1
//         },

//         {
//             fieldname: "company",
//             label: __("Company"),
//             fieldtype: "Link",
//             options: "Company",
//             default: frappe.defaults.get_default("company")
//         },

//         {
//             fieldname: "from_date",
//             label: __("From Date"),
//             fieldtype: "Date",
//             reqd: 1,
//             default: frappe.datetime.add_months(frappe.datetime.get_today(), -1)
//         },

//         {
//             fieldname: "to_date",
//             label: __("To Date"),
//             fieldtype: "Date",
//             reqd: 1,
//             default: frappe.datetime.get_today()
//         },

//         {
//             fieldname: "item_group",
//             label: __("Item Group"),
//             fieldtype: "Link",
//             options: "Item Group"
//         },

//         {
//             fieldname: "item_code",
//             label: __("Item"),
//             fieldtype: "Link",
//             options: "Item"
//         }
//     ],

//     formatter: function(value, row, column, data, default_formatter) {

//         value = default_formatter(value, row, column, data);

//         // Skip fixed columns
//         let fixed_columns = [
//             "user",
//             "item_code",
//             "item_name"
//         ];

//         if (!fixed_columns.includes(column.fieldname) && data) {

//             let qty = data[column.fieldname];

//             if (qty > 0) {

//                 value = `<span style="
//                     color: #058203;
//                     // font-weight: bold;
//                 ">${value}</span>`;

//             } else if (qty < 0) {

//                 value = `<span style="
//                     color: #ff0000;
//                     // font-weight: bold;
//                 ">${value}</span>`;

//             } else if (qty == 0) {

//                 value = `<span style="
//                     color: #ffda1f;
//                     // font-weight: bold;
//                 ">${value}</span>`;
//             }
//         }

//         return value;
//     }
// };






































// frappe.query_reports["User Warehouse Stock Balance"] = {

//     onload: function(report) {

//         report.set_filter_value("warehouse", [

//             "New store - NAVYA",
//             "Default Transit - NAVYA",
//             "Kit Store - NAVYA",
//             "Navya Store Office - NAVYA",
//             "Purchase Station - NAVYA"

//         ]);

//     },

//     filters: [

//         {
//             fieldname: "user",
//             label: __("User"),
//             fieldtype: "Link",
//             options: "User",
//             reqd: 1
//         },

//         {
//             fieldname: "warehouse",
//             label: __("Warehouses"),
//             fieldtype: "MultiSelectList",
//             width: "300",

//             get_data: function(txt) {

//                 return frappe.db.get_link_options(
//                     "Warehouse",
//                     txt,
//                     {}
//                 );

//             }
//         },

//         {
//             fieldname: "company",
//             label: __("Company"),
//             fieldtype: "Link",
//             options: "Company",
//             default: frappe.defaults.get_default("company")
//         },

//         {
//             fieldname: "from_date",
//             label: __("From Date"),
//             fieldtype: "Date",
//             reqd: 1,
//             default: frappe.datetime.add_months(
//                 frappe.datetime.get_today(),
//                 -1
//             )
//         },

//         {
//             fieldname: "to_date",
//             label: __("To Date"),
//             fieldtype: "Date",
//             reqd: 1,
//             default: frappe.datetime.get_today()
//         },

//         {
//             fieldname: "item_group",
//             label: __("Item Group"),
//             fieldtype: "Link",
//             options: "Item Group"
//         },

//         {
//             fieldname: "item_code",
//             label: __("Item"),
//             fieldtype: "Link",
//             options: "Item"
//         }
//     ],

//     formatter: function(
//         value,
//         row,
//         column,
//         data,
//         default_formatter
//     ) {

//         value = default_formatter(
//             value,
//             row,
//             column,
//             data
//         );

//         // Fixed columns
//         let fixed_columns = [

//             "user",
//             "item_code",
//             "item_name"

//         ];

//         if (
//             !fixed_columns.includes(column.fieldname)
//             && data
//         ) {

//             let qty = data[column.fieldname];

//             if (qty > 0) {

//                 value = `
//                     <span style="color:#058203;">
//                         ${value}
//                     </span>
//                 `;

//             } else if (qty < 0) {

//                 value = `
//                     <span style="color:#ff0000;">
//                         ${value}
//                     </span>
//                 `;

//             } else if (qty == 0) {

//                 value = `
//                     <span style="color:#ffda1f;">
//                         ${value}
//                     </span>
//                 `;
//             }
//         }

//         return value;
//     }
// };








































frappe.query_reports["User Warehouse Stock Balance"] = {

    onload: function(report) {

        let user_filter = report.get_filter("user");

        user_filter.$input.on("change", function() {

            let user = report.get_filter_value("user");

            if (!user) {
                report.set_filter_value("warehouse", []);
                return;
            }

            frappe.call({
                method: "fashion_navya.fashion_navya.report.user_warehouse_stock_balance.user_warehouse_stock_balance.get_user_warehouses",
                args: {
                    user: user
                },
                callback: function(r) {

                    if (r.message) {

                        report.set_filter_value(
                            "warehouse",
                            r.message
                        );

                    }
                }
            });

        });

    },

    filters: [

        {
            fieldname: "user",
            label: __("User"),
            fieldtype: "Link",
            options: "User",
            reqd: 1
        },

        {
            fieldname: "warehouse",
            label: __("Warehouses"),
            fieldtype: "MultiSelectList",
            width: 300,

            get_data: function(txt) {

                return frappe.db.get_link_options(
                    "Warehouse",
                    txt
                );

            }
        },

        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            default: frappe.defaults.get_default("company")
        },

        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.add_months(
                frappe.datetime.get_today(),
                -1
            )
        },

        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.get_today()
        },

        {
            fieldname: "item_group",
            label: __("Item Group"),
            fieldtype: "Link",
            options: "Item Group"
        },

        {
            fieldname: "item_code",
            label: __("Item"),
            fieldtype: "Link",
            options: "Item"
        }

    ],

    formatter: function(
        value,
        row,
        column,
        data,
        default_formatter
    ) {

        value = default_formatter(
            value,
            row,
            column,
            data
        );

        let fixed_columns = [
            "user",
            "item_code",
            "item_name"
        ];

        if (
            !fixed_columns.includes(column.fieldname)
            && data
        ) {

            let qty = data[column.fieldname];

            if (qty > 0) {

                value = `
                    <span style="color:#058203;">
                        ${value}
                    </span>
                `;

            } else if (qty < 0) {

                value = `
                    <span style="color:#ff0000;">
                        ${value}
                    </span>
                `;

            } else if (qty == 0) {

                value = `
                    <span style="color:#ffda1f;">
                        ${value}
                    </span>
                `;

            }
        }

        return value;
    }
};




















